### PR TITLE
fix: [SIW-1795] Problem with SPID Login Using Custom Tab

### PR DESCRIPTION
**Pull request by RiccardoMolinari95**
#### List of Changes

- Renamed the screen `PidSpidLoginScreen` to `PidSpidIdpSelectionScreen`. When an IDP is selected, the authorization flow starts, and the user is redirected to the new `PidSpidLoginScreen` after obtaining the `authUrl`.

- Created the new screen `PidSpidLoginScreen`, which uses a `WebView` to load `authUrl`. Once the authorization is completed and the redirect URL is obtained, the flow continues to complete the PID issuance.

- Updated the `pid.ts` file to manage authentication via SPID and CIE L3, including the creation of two new thunks: `preparePidFlowParamsThunk` and `continuePidFlowThunk`. `preparePidFlowParamsThunk` obtains the `authUrl`, passes it to the respective `WebView` for rendering, and saves the necessary parameters for the next step in the state. `continuePidFlowThunk` receives the redirect URL from the WebView and continues the process to obtain the PID.

- The parameters PreparePidFlowParamsThunkInput and PreparePidFlowParamsThunkOutput have been set to handle both the CIE and SPID flows, so some parameters can be optional, such as ciePin.

-  Added `openUrlAndListenForAuthRedirect`, which opens the `authUrl` in the browser for CIE ID and provides the `authRedirectURL`. This step was previously handled by `CompleteUserAuthorizationWithQueryMode`

- Added the function `buildAuthorizationUrl` in step `04-complete-user-authorization`, which constructs the `authUrl` that will be passed to the WebView for CIE L3 and SPID, and to `openUrlAndListenForAuthRedirect` for CIE ID

- The parameters `issuerRequestUri`, `clientId`, `issuerConf`, and `idpHint` have been removed from the type `CompleteUserAuthorizationWithQueryMode`. Now `CompleteUserAuthorizationWithQueryMode` receives the `authRedirectURL` from the WebView for CIE L3 or SPID, or from the browser redirect for CIE L2, and performs the parsing.

#### Motivation and Context

This PR removes the usage of the `io-react-native-login-utils` library and its `Custom Tab `implementation for SPID authentication to obtain the PID. The `Custom Tab` has been replaced with a `WebView` to avoid issues that some users were experiencing during the SPID login process.
The login flow via SPID has been unified with the CIE L3 flow due to their similarities and has been split into two parts. The CIE L2 flow remains the same as before.

#### How Has This Been Tested?

Successfully obtained the PID using all three authentication methods: SPID, CIE L2, and CIE L3.

#### Screenshots (if appropriate):

https://github.com/user-attachments/assets/ba5af9fb-20da-4184-83c4-8cd0e85b50eb

#### Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by LazyAfternoons — CHANGES_REQUESTED** [review]
Good work, I'd like to point out a few things we might refine, if that makes sense to you: 
- `startUserAuthroization` should keep the same implementation as before to keep it consistent between eID and a generic credential. This allow us to have a strict signature in terms of input/output without the need of declaring the `authUrl` which might not be needed in case of a generic credential; 
- We should create a new function in the `04-complete-user-authroization` step which builds the authentication urls and will be only used in case of a eID issuance, similarly 
to `getCredentialToBePresented` which is used only for other credentials; 
- In the `completeUserAuthorizationWithQueryMode` we should only parse the result redirect url which comes from the consumer who performed the strong authentication; 
- We should move the CieID logic to the consumer app since we don't manage any kind of authentication directly in the library anymore.

Let me know what you think. 

```mermaid
graph TD;
    2[...]
    3[startUserAuthorization]
    C4[getRequestedCredentialToBePresented]
    C4.1[completeUserAuthorizationWithFormPostJwtMode]
    E4[getAuthorizationUrl]
    E4.1[Consumer performs strong authentication via SPID/CIE with authroization URL]
    E4.2[completeUserAuthorizationWithQueryMode]
    5[authorizeAccess]
    6[..]
    credSel{Is credential an eID?}
    
    2 --> 3
    3 --> credSel
    credSel -->|Yes| E4
    credSel -->|No| C4
    C4 --> C4.1
    C4.1 --> 5
    E4 --> E4.1
    E4.1 --> E4.2
    E4.2 --> 5
    5 --> 6
```

**Comment by RiccardoMolinari95** [comment]
> Good work, I'd like to point out a few things we might refine, if that makes sense to you:
> 
> * `startUserAuthroization` should keep the same implementation as before to keep it consistent between eID and a generic credential. This allow us to have a strict signature in terms of input/output without the need of declaring the `authUrl` which might not be needed in case of a generic credential;
> * We should create a new function in the `04-complete-user-authroization` step which builds the authentication urls and will be only used in case of a eID issuance, similarly
>   to `getCredentialToBePresented` which is used only for other credentials;
> * In `completeUserAuthorizationWithQueryMode` we should only parse the result redirect url which comes from the consumer who performed the strong authentication;
> * We should move the CieID logic to the consumer app since we don't manage any kind of authentication directly in the library anymore.
> 
> Let me know what you think.
> 
> ```mermaid
> graph TD;
>     2[...]
>     3[startUserAuthorization]
>     C4[getRequestedCredentialToBePresented]
>     C4.1[completeUserAuthorizationWithFormPostJwtMode]
>     E4[getAuthorizationUrl]
>     E4.1[Consumer performs strong authentication via SPID/CIE with authroization URL]
>     E4.2[completeUserAuthorizationWithQueryMode]
>     5[authorizeAccess]
>     6[..]
>     credSel{Is credential an eID?}
>     
>     2 --> 3
>     3 --> credSel
>     credSel -->|Yes| E4
>     credSel -->|No| C4
>     C4 --> C4.1
>     C4.1 --> 5
>     E4 --> E4.1
>     E4.1 --> E4.2
>     E4.2 --> 5
>     5 --> 6
> ```

Everything makes sense to me! I also prefer not to use optional parameters.
I will proceed with making the changes.

File: src/index.ts
Comment (LazyAfternoons) [inline src/index.ts:15]: Do we need this export?

Comment (RiccardoMolinari95) [reply src/index.ts:15]: Yes, it makes sense

File: src/credential/issuance/__tests__/04-complete-user-authorization.test.ts
Comment (gispada) [inline src/credential/issuance/__tests__/04-complete-user-authorization.test.ts:18]: Is there a reason we are not testing `completeUserAuthorizationWithQueryMode` but only `parseAuthroizationResponse`?

File: src/credential/issuance/__tests__/04-complete-user-authorization.test.ts
Comment (gispada) [inline src/credential/issuance/__tests__/04-complete-user-authorization.test.ts:34]: Same as above.

File: src/credential/issuance/__tests__/04-complete-user-authorization.test.ts
Comment (gispada) [inline src/credential/issuance/__tests__/04-complete-user-authorization.test.ts:50]: Same as above.

File: example/ios/IoReactNativeWalletExample.xcodeproj/project.pbxproj
Comment (gispada) [inline example/ios/IoReactNativeWalletExample.xcodeproj/project.pbxproj:1]: Can you check why this file was modified? Usually this file should not change.

File: example/src/thunks/pidCieID.ts
Comment (gispada) [inline example/src/thunks/pidCieID.ts:1]: Do we really need to have a separate flow for CieID? Would it be possible to handle all three scenarios the same way?
1. `preparePidFlowParamsThunk`
2. Open webview for Spid and CIE L3 / open browser for CieID -> get the auth redirect url
4. `continuePidFlowThunk`

File: src/credential/issuance/__tests__/04-complete-user-authorization.test.ts
Comment (RiccardoMolinari95) [reply src/credential/issuance/__tests__/04-complete-user-authorization.test.ts:50]: It is an oversight, thank you for noticing. I will correct it

File: src/credential/issuance/04-complete-user-authorization.ts
Comment (RiccardoMolinari95) [reply src/credential/issuance/04-complete-user-authorization.ts:61]: It is an oversight, thank you for noticing. I will correct it

File: example/src/thunks/pidCieID.ts
Comment (RiccardoMolinari95) [reply example/src/thunks/pidCieID.ts:1]: Yes could be an option, for future improvement

File: src/credential/issuance/04-complete-user-authorization.ts
Comment (gispada) [inline src/credential/issuance/04-complete-user-authorization.ts:61]:
```suggestion
 * WARNING: This function must be called after {@link startUserAuthorization}. The generated authUrl must be used to open a browser or webview capable of catching the redirectSchema to perform a get request to the authorization endpoint.
```

File: example/src/thunks/pidCieID.ts
Comment (gispada) [reply example/src/thunks/pidCieID.ts:1]: Can you create a Jira Task and add a TODO comment to reference that task? We can treat it as a low priority task, but I would like to remove code duplication, even if it's just an example app.

Comment (RiccardoMolinari95) [reply example/src/thunks/pidCieID.ts:1]: Good idea, so we can keep track of the activity

File: example/src/store/reducers/pid.ts
Comment (RiccardoMolinari95) [reply example/src/store/reducers/pid.ts:173]: I set "spid" as the default value in case neither `cieL3IsLoading` nor `spidIsLoading` is true. This ensures that `authMethod` always has a valid value, even if neither of the authentication methods is currently loading.

**Review by gispada — APPROVED** [review]
LGTM. I'd wait for another approval though, as this PR involves a high risk flow.

**Review by grausof — APPROVED** [review]
LGTM 🎉

**Comment by mastro993** [comment]
On iOS, I'm unable to complete the PID issuance with SPID. After a successful login, I get stuck on this screen:
 


Am I missing something?

**Comment by RiccardoMolinari95** [comment]
> On iOS, I'm unable to complete the PID issuance with SPID. After a successful login, I get stuck on this screen:
> 
> 
> Am I missing something?

Is this a mistake that you always have ? Have you tried deleting the example app from the phone and restarting from a clean situation ?

File: example/src/screens/login/PidLoginScreen.tsx
Comment (mastro993) [inline example/src/screens/login/PidLoginScreen.tsx:54]: `includes` it's too broad, we could use `startsWith` instead. What do you think?

```suggestion
    if (url.startsWith("iowallet")) {
```

Comment (RiccardoMolinari95) [reply example/src/screens/login/PidLoginScreen.tsx:54]: Thanks for the suggestions. i will implement them

File: example/src/thunks/pid.ts
Comment (grausof) [inline example/src/thunks/pid.ts:68]: Always use camel case

File: example/src/thunks/pid.ts
Comment (grausof) [inline example/src/thunks/pid.ts:109]:
```suggestion
  const redirectUri = isCie ? CIE_L3_REDIRECT_URI : REDIRECT_URI;
```

File: example/src/screens/login/PidLoginScreen.tsx
Comment (grausof) [inline example/src/screens/login/PidLoginScreen.tsx:58]: Avoid reading from env. You can get the redirectUri from `route.params`

File: example/src/screens/login/PidSpidIdpSelectionScreen.tsx
Comment (grausof) [inline example/src/screens/login/PidSpidIdpSelectionScreen.tsx:40]: here you can also pass redirectUri since you need it in WebView

**Review by grausof — CHANGES_REQUESTED** [review]

File: example/src/screens/login/PidSpidIdpSelectionScreen.tsx
Comment (RiccardoMolinari95) [reply example/src/screens/login/PidSpidIdpSelectionScreen.tsx:40]: Thanks, nice suggestions

File: example/src/screens/login/PidLoginScreen.tsx
Comment (grausof) [inline example/src/screens/login/PidLoginScreen.tsx:84]: What do you think about integrating this component into an ad-hoc io-react-native-wallet component like done for CieL3 WebView?

**Review by grausof — APPROVED** [review]
Great 🎉

**Review by mastro993 — APPROVED** [review]
LGTM

**Comment by grausof** [comment]
This PR is ok for now. Need a separate PR to fix the sample app on iOS!

**Pull request merged**
